File: src/store/rustore.ts
import { existsSync, openAsBlob } from "node:fs";
import { basename } from "node:path";
import { createSign } from "crypto";
import { z } from "zod";
import type { StoreClient, UploadResult } from "./store-client.js";
import { AbstractStoreClient } from "./base-client.js";
import { validatePackageName } from "../utils/sanitize.js";

const BASE = "https://public-api.rustore.ru/public/v1";
const AUTH_URL = "https://public-api.rustore.ru/public/auth";

interface RuStoreCredentials {
  companyId: string;
  keyId: string;
  privateKey: string;
}

interface TokenCache {
  token: string;
  expiresAt: number;
}


interface ReleaseNoteEntry {
  language: string;
  text: string;
}

interface DraftState {
  versionId: number;
  releaseNotes: ReleaseNoteEntry[];
}
const ruStoreCodeSchema = z.string().min(1).max(128).regex(/^[A-Za-z0-9._-]+$/);
const ruStoreRecordSchema = z.record(z.string(), z.unknown());
const ruStoreResponseSchema = z.object({
  code: ruStoreCodeSchema,
  message: z.string().max(4096).optional(),
  body: z.unknown().optional(),
}).passthrough();
const ruStoreCredentialsSchema = z.object({
  companyId: z.string().min(1).max(256),
  keyId: z.string().min(1).max(256),
  privateKey: z.string().min(1).max(1024 * 1024),
}).strict();
const ruStoreAuthBodySchema = z.object({
  jwtToken: z.string().min(1).max(64 * 1024),
  ttl: z.number().finite().positive().max(86_400),
}).passthrough();
const ruStoreDraftBodySchema = z.object({
  versionId: z.number().int().safe().positive(),
}).passthrough();
const ruStoreVersionSchema = z.object({
  versionId: z.number().int().safe(),
  versionCode: z.number().int().safe().optional(),
  versionName: z.string().max(256).optional(),
  appStatus: z.string().max(128).optional(),
  publishType: z.string().max(128).optional(),
}).passthrough();
const ruStoreVersionListSchema = z.array(ruStoreVersionSchema).max(10_000);

function responseRecord(value: unknown): Record<string, unknown> {
  const result = ruStoreRecordSchema.safeParse(value);
  if (!result.success) {
    throw new Error("RuStore returned an invalid response.");
  }
  return result.data;
}

function responseCode(value: unknown): { record: Record<string, unknown>; code: string } {
  const result = ruStoreResponseSchema.safeParse(value);
  if (!result.success) {
    throw new Error("RuStore returned an invalid response.");
  }
  return { record: result.data, code: result.data.code };
}

function loadCredentials(): RuStoreCredentials {
  const keyJson = process.env.RUSTORE_KEY_JSON;
  // Try JSON config first
  if (keyJson) {
    if (Buffer.byteLength(keyJson, "utf8") > 1024 * 1024) {
      throw new Error("RuStore: RUSTORE_KEY_JSON exceeds the size limit");
    }
    try {
      const result = ruStoreCredentialsSchema.safeParse(JSON.parse(keyJson));
      if (!result.success) throw new Error("invalid credentials");
      return result.data;
    } catch {
      throw new Error("RuStore: RUSTORE_KEY_JSON is not valid credentials JSON");
    }
  }

  // Try individual env vars
  const companyId = process.env.RUSTORE_COMPANY_ID;
  const keyId = process.env.RUSTORE_KEY_ID;
  const privateKey = process.env.RUSTORE_PRIVATE_KEY;

  const result = ruStoreCredentialsSchema.safeParse({ companyId, keyId, privateKey });
  if (result.success) {
    return result.data;
  }

  throw new Error(
    "RuStore: missing credentials. Set RUSTORE_KEY_JSON or (RUSTORE_COMPANY_ID + RUSTORE_KEY_ID + RUSTORE_PRIVATE_KEY) environment variables."
  );
}

function createJwt(credentials: RuStoreCredentials): string {
  const header = Buffer.from(JSON.stringify({ alg: "RS256", typ: "JWT" })).toString("base64url");
  const payload = Buffer.from(
    JSON.stringify({ keyId: credentials.keyId, timestamp: Date.now() })
  ).toString("base64url");

  const signingInput = `${header}.${payload}`;
  const sign = createSign("RSA-SHA256");
  sign.update(signingInput);
  const signature = sign.sign(credentials.privateKey, "base64url");

  return `${signingInput}.${signature}`;
}

export class RuStoreClient extends AbstractStoreClient implements StoreClient {
  private tokenCache: TokenCache | null = null;
  private drafts = new Map<string, DraftState>();

  protected get apiErrorPrefix(): string {
    return "RuStore API";
  }

  /** RuStore uses "Public-Token" header instead of "Authorization: Bearer" */
  protected override authHeader(token: string): Record<string, string> {
    return { "Public-Token": token };
  }

  // ── Auth ─────────────────────────────────────────────────────────────────────

  private async getToken(): Promise<string> {
    const now = Date.now();
    // Refresh 60 seconds before expiry
    if (this.tokenCache && this.tokenCache.expiresAt > now + 60_000) {
      return this.tokenCache.token;
    }

    const credentials = loadCredentials();
    const jwtToken = createJwt(credentials);

    const res = await this.fetchWithTimeout(AUTH_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jwtToken }),
    });

    if (!res.ok) {
      await res.body?.cancel();
      throw new Error(`RuStore auth failed with HTTP ${res.status}.`);
    }

    const { record: data, code } = responseCode(await this.readJson(res));
    if (code !== "OK") throw new Error(`RuStore auth failed with code ${code}.`);
    const authResult = ruStoreAuthBodySchema.safeParse(data.body);
    if (!authResult.success) {
      throw new Error("RuStore returned invalid authentication data.");
    }
    const authBody = authResult.data;
    this.tokenCache = {
      token: authBody.jwtToken,
      expiresAt: Date.now() + authBody.ttl * 1000,
    };
    return this.tokenCache.token;
  }

  private async createDraftVersion(packageName: string, token: string): Promise<number> {
    const { record: data, code } = responseCode(await this.api(
      "POST",
      `${BASE}/application/${encodeURIComponent(packageName)}/version`,
      token,
      { whatsNew: {} }
    ));
    if (code !== "OK") throw new Error(`RuStore draft creation failed with code ${code}.`);
    const bodyResult = ruStoreDraftBodySchema.safeParse(data.body);
    if (!bodyResult.success) {
      throw new Error("RuStore returned an invalid draft version ID.");
    }
    return bodyResult.data.versionId;
  }

  // ── Public API ───────────────────────────────────────────────────────────────

  async upload(packageName: string, filePath: string): Promise<UploadResult> {
    validatePackageName(packageName);
    if (!existsSync(filePath)) {
      throw new Error(`File not found: ${filePath}`);
    }

    const token = await this.getToken();
    const versionId = await this.createDraftVersion(packageName, token);

    const isAab = filePath.toLowerCase().endsWith(".aab");
    const uploadPath = isAab ? "aab" : "apk";
    const fileName = basename(filePath);

    const formData = new FormData();
    const blob = await openAsBlob(filePath, { type: "application/octet-stream" });
    formData.append("file", blob, fileName);

    const uploadUrl =
      `${BASE}/application/${encodeURIComponent(packageName)}/version/${versionId}/${uploadPath}` +
      `?servicesType=Unknown&isMainApk=true`;

    const res = await this.fetchWithTimeout(uploadUrl, {
      method: "POST",
      headers: {
        "Public-Token": token,
        // No Content-Type — let fetch set multipart boundary automatically
      },
      body: formData,
    }, 10 * 60_000);

    if (!res.ok) {
      await res.body?.cancel();
      await this.deleteDraft(packageName, versionId, token);
      throw new Error(`RuStore upload failed with HTTP ${res.status}.`);
    }

    const { code } = responseCode(await this.readJson(res));
    if (code !== "OK") {
      await this.deleteDraft(packageName, versionId, token);
      throw new Error(`RuStore upload failed with code ${code}.`);
    }

    this.drafts.set(packageName, { versionId, releaseNotes: [] });

    return { versionId: String(versionId) };
  }

  async setReleaseNotes(packageName: string, language: string, text: string): Promise<void> {
    validatePackageName(packageName);
    if (!/^[A-Za-z]{2,3}(?:-[A-Za-z0-9]{2,8})*$/.test(language)) {
      throw new Error("Invalid RuStore release language.");
    }
    if (Buffer.byteLength(text, "utf8") > 50 * 1024 || text.includes("\0")) {
      throw new Error("RuStore release notes exceed the size limit.");
    }
    const draft = this.drafts.get(packageName);
    if (!draft) {
      throw new Error(`RuStore: no active upload for "${packageName}". Call rustore_upload first.`);
    }
    const idx = draft.releaseNotes.findIndex(n => n.language === language);
    if (idx >= 0) {
      draft.releaseNotes[idx].text = text;
    } else {
      draft.releaseNotes.push({ language, text });
    }
  }

  async submit(packageName: string, _options?: { rollout?: number }): Promise<void> {
    validatePackageName(packageName);
    if (
      _options?.rollout !== undefined
      && (!Number.isFinite(_options.rollout) || _options.rollout !== 1)
    ) {
      throw new Error("RuStore does not support staged rollout.");
    }
    const draft = this.drafts.get(packageName);
    if (!draft) {
      throw new Error(`RuStore: no active upload for "${packageName}". Call rustore_upload first.`);
    }

    const token = await this.getToken();
    const { versionId, releaseNotes } = draft;

    // If there are release notes — patch them first
    if (releaseNotes.length > 0) {
      const whatsNew: Record<string, string> = {};
      for (const note of releaseNotes) {
        whatsNew[note.language] = note.text;
      }
      const patchData = await this.api(
        "PATCH",
        `${BASE}/application/${encodeURIComponent(packageName)}/version/${versionId}/publishing-settings`,
        token,
        { whatsNew }
      );
      const { code } = responseCode(patchData);
      if (code !== "OK") {
        throw new Error(`RuStore release notes update failed with code ${code}.`);
      }
    }

    // Submit for moderation
    const submitData = await this.api(
      "POST",
      `${BASE}/application/${encodeURIComponent(packageName)}/version/${versionId}/submit-for-moderation`,
      token
    );

    const { code } = responseCode(submitData);
    if (code !== "OK") {
      throw new Error(`RuStore submission failed with code ${code}.`);
    }

    this.drafts.delete(packageName);
  }

  async getReleases(packageName: string): Promise<string> {
    validatePackageName(packageName);
    const token = await this.getToken();

    const data = await this.api(
      "GET",
      `${BASE}/application/${encodeURIComponent(packageName)}/version`,
      token
    );

    const { record, code } = responseCode(data);
    if (code !== "OK") {
      throw new Error(`RuStore release listing failed with code ${code}.`);
    }
    const versionsResult = ruStoreVersionListSchema.safeParse(record.body);
    if (!versionsResult.success) {
      throw new Error("RuStore returned an invalid release list.");
    }

    if (versionsResult.data.length === 0) return `${packageName}: no versions found`;
    const lines = versionsResult.data.map((version) => {
      const versionName = version.versionName
        ?.replace(/[\u0000-\u001f\u007f]/g, "")
        .slice(0, 256);
      const versionCode = version.versionCode ?? "?";
      const status = version.appStatus
        ?.replace(/[\u0000-\u001f\u007f]/g, "")
        .slice(0, 128) ?? "unknown";
      const displayVersion = versionName
        ? `${versionName} (${versionCode})`
        : `versionId=${String(version.versionId)}`;
      return `  v${displayVersion} — ${status}`;
    });
    return `${packageName}:\n${lines.join("\n")}`;
  }

  async discard(packageName: string): Promise<void> {
    validatePackageName(packageName);
    const draft = this.drafts.get(packageName);
    if (!draft) {
      throw new Error(`RuStore: no active draft for "${packageName}"`);
    }

    const token = await this.getToken();
    await this.deleteDraft(packageName, draft.versionId, token);
    this.drafts.delete(packageName);
  }

  // ── Private helpers ──────────────────────────────────────────────────────────

  private async deleteDraft(packageName: string, versionId: number, token: string): Promise<void> {
    const res = await this.fetchWithTimeout(
      `${BASE}/application/${encodeURIComponent(packageName)}/version/${versionId}`,
      {
        method: "DELETE",
        headers: { "Public-Token": token },
      }
    );
    // Best-effort cleanup — do not throw
    if (!res.ok) {
      console.error(`RuStore: failed to delete draft version ${versionId} (${res.status})`);
    }
  }
}

File: cli/src/store/rustore.rs
//! RuStore management.
//! Auth: RSA RS256 JWT → RuStore token.
//! Env: RUSTORE_KEY_JSON  or  RUSTORE_COMPANY_ID + RUSTORE_KEY_ID + RUSTORE_PRIVATE_KEY

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use serde_json::json;

use super::{draft, jwt};

const BASE: &str = "https://public-api.rustore.ru/public/v1";
const AUTH_URL: &str = "https://public-api.rustore.ru/public/auth";

// ── Types ────────────────────────────────────────────────────────────────────

#[derive(Deserialize)]
struct Credentials {
    #[serde(rename = "companyId")]
    company_id: String,
    #[serde(rename = "keyId")]
    key_id: String,
    #[serde(rename = "privateKey")]
    private_key: String,
}

#[derive(Serialize, Deserialize)]
struct DraftState {
    version_id: i64,
    release_notes: Vec<Note>,
}

#[derive(Serialize, Deserialize, Clone)]
struct Note {
    language: String,
    text: String,
}

// ── Auth ─────────────────────────────────────────────────────────────────────

fn load_credentials() -> Result<Credentials> {
    // Prefer JSON bundle
    if let Ok(raw) = std::env::var("RUSTORE_KEY_JSON") {
        return serde_json::from_str(&raw).context("RUSTORE_KEY_JSON is not valid JSON");
    }
    // Fall back to individual vars
    match (
        std::env::var("RUSTORE_COMPANY_ID").ok(),
        std::env::var("RUSTORE_KEY_ID").ok(),
        std::env::var("RUSTORE_PRIVATE_KEY").ok(),
    ) {
        (Some(company_id), Some(key_id), Some(private_key)) => {
            Ok(Credentials { company_id, key_id, private_key })
        }
        _ => anyhow::bail!(
            "RuStore: missing credentials.\n\
             Set RUSTORE_KEY_JSON  or  RUSTORE_COMPANY_ID + RUSTORE_KEY_ID + RUSTORE_PRIVATE_KEY."
        ),
    }
}

fn get_token() -> Result<String> {
    let creds = load_credentials()?;
    let payload = json!({
        "keyId": creds.key_id,
        "timestamp": jwt::now_millis(),
    });
    let jwt_token = jwt::create_rs256(&creds.private_key, &payload)?;

    let resp: serde_json::Value = client()
        .post(AUTH_URL)
        .json(&json!({ "jwtToken": jwt_token }))
        .send()
        .context("Failed to connect to RuStore auth endpoint")?
        .error_for_status()
        .context("RuStore auth returned error")?
        .json()
        .context("Failed to parse RuStore auth response")?;

    if resp["code"].as_str() != Some("OK") {
        let msg = resp["message"]
            .as_str()
            .or_else(|| resp["code"].as_str())
            .unwrap_or("unknown");
        anyhow::bail!("RuStore auth error: {}", msg);
    }

    resp["body"]["jwtToken"]
        .as_str()
        .map(|s| s.to_string())
        .context("RuStore auth: no jwtToken in response body")
}

// ── HTTP helpers ─────────────────────────────────────────────────────────────

fn client() -> reqwest::blocking::Client {
    reqwest::blocking::Client::new()
}

fn api(
    method: &str,
    url: &str,
    token: &str,
    body: Option<&serde_json::Value>,
) -> Result<serde_json::Value> {
    let c = client();
    let mut req = match method {
        "GET" => c.get(url),
        "POST" => c.post(url),
        "PUT" => c.put(url),
        "PATCH" => c.patch(url),
        "DELETE" => c.delete(url),
        other => anyhow::bail!("Unknown HTTP method: {}", other),
    };
    req = req
        .header("Public-Token", token)
        .header("Content-Type", "application/json");
    if let Some(b) = body {
        req = req.json(b);
    }
    let resp = req.send().with_context(|| format!("{} {}", method, url))?;
    if resp.status().as_u16() == 204 {
        return Ok(json!({}));
    }
    resp.error_for_status()
        .with_context(|| format!("{} {} failed", method, url))?
        .json()
        .context("Failed to parse JSON response")
}

fn create_draft_version(package: &str, token: &str) -> Result<i64> {
    let resp = api(
        "POST",
        &format!("{}/application/{}/version", BASE, urlenc(package)),
        token,
        Some(&json!({ "whatsNew": {} })),
    )?;
    check_ok(&resp, "Failed to create draft version")?;
    resp["body"]["versionId"]
        .as_i64()
        .context("RuStore: no versionId in create version response")
}

fn delete_version(package: &str, version_id: i64, token: &str) {
    let url = format!("{}/application/{}/version/{}", BASE, urlenc(package), version_id);
    let _ = client().delete(&url).header("Public-Token", token).send();
}

fn check_ok(resp: &serde_json::Value, context: &str) -> Result<()> {
    if resp["code"].as_str() != Some("OK") {
        let msg = resp["message"]
            .as_str()
            .or_else(|| resp["code"].as_str())
            .unwrap_or("unknown");
        anyhow::bail!("{}: {}", context, msg);
    }
    Ok(())
}

fn urlenc(s: &str) -> String {
    s.replace('/', "%2F")
}

// ── Public API ────────────────────────────────────────────────────────────────

pub fn upload(package: &str, file_path: &str) -> Result<()> {
    if !std::path::Path::new(file_path).exists() {
        anyhow::bail!("File not found: {}", file_path);
    }

    let token = get_token()?;
    let version_id = create_draft_version(package, &token)?;

    let is_aab = file_path.to_lowercase().ends_with(".aab");
    let upload_type = if is_aab { "aab" } else { "apk" };
    let file_name = std::path::Path::new(file_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    let file_part = reqwest::blocking::multipart::Part::file(file_path)
        .context("Failed to open file for upload")?
        .file_name(file_name)
        .mime_str("application/octet-stream")?;
    let form = reqwest::blocking::multipart::Form::new().part("file", file_part);

    let upload_url = format!(
        "{}/application/{}/version/{}/{}?servicesType=Unknown&isMainApk=true",
        BASE,
        urlenc(package),
        version_id,
        upload_type
    );

    let resp = client()
        .post(&upload_url)
        .header("Public-Token", &token)
        .multipart(form)
        .send()
        .context("Failed to upload to RuStore")?;

    if !resp.status().is_success() {
        let status = resp.status();
        let text = resp.text().unwrap_or_default();
        delete_version(package, version_id, &token);
        anyhow::bail!("RuStore: upload failed {}: {}", status, text);
    }

    let data: serde_json::Value = resp.json().context("Failed to parse RuStore upload response")?;
    if data["code"].as_str() != Some("OK") {
        let msg = data["message"].as_str().unwrap_or("unknown");
        delete_version(package, version_id, &token);
        anyhow::bail!("RuStore: upload error: {}", msg);
    }

    draft::save(
        "rustore",
        package,
        &DraftState { version_id, release_notes: vec![] },
    )?;

    println!(
        "Uploaded to RuStore. Version ID: {}\nDraft open — call 'rustore set-notes' and 'rustore submit' to send for moderation.",
        version_id
    );
    Ok(())
}

pub fn set_notes(package: &str, language: &str, text: &str) -> Result<()> {
    if text.len() > 500 {
        anyhow::bail!("What's new text exceeds 500 characters ({})", text.len());
    }
    let mut state: DraftState = draft::load("rustore", package)?;
    upsert_note(&mut state.release_notes, language, text);
    draft::save("rustore", package, &state)?;
    println!("RuStore what's new set for {} ({}/500 chars)", language, text.len());
    Ok(())
}

pub fn submit(package: &str) -> Result<()> {
    let state: DraftState = draft::load("rustore", package)?;
    let token = get_token()?;

    // Patch release notes if any
    if !state.release_notes.is_empty() {
        let mut whats_new = serde_json::Map::new();
        for note in &state.release_notes {
            whats_new.insert(note.language.clone(), json!(note.text));
        }
        let patch = api(
            "PATCH",
            &format!(
                "{}/application/{}/version/{}/publishing-settings",
                BASE, urlenc(package), state.version_id
            ),
            &token,
            Some(&json!({ "whatsNew": whats_new })),
        )?;
        check_ok(&patch, "Failed to set release notes")?;
    }

    // Submit for moderation
    let submit_data = api(
        "POST",
        &format!(
            "{}/application/{}/version/{}/submit-for-moderation",
            BASE, urlenc(package), state.version_id
        ),
        &token,
        None,
    )?;
    check_ok(&submit_data, "Submit failed")?;

    draft::delete("rustore", package);
    println!(
        "Submitted to RuStore for moderation: {}\nPublication requires moderation approval (typically 1–3 business days).",
        package
    );
    Ok(())
}

pub fn get_versions(package: &str) -> Result<()> {
    let token = get_token()?;
    let data = api(
        "GET",
        &format!("{}/application/{}/version", BASE, urlenc(package)),
        &token,
        None,
    )?;
    check_ok(&data, "getReleases failed")?;

    let versions = data["body"].as_array().map(|v| v.as_slice()).unwrap_or(&[]);
    if versions.is_empty() {
        println!("{}: no versions found", package);
        return Ok(());
    }

    let lines: Vec<String> = versions
        .iter()
        .map(|v| {
            let ver = if let (Some(name), Some(code)) =
                (v["versionName"].as_str(), v["versionCode"].as_i64())
            {
                format!("{} ({})", name, code)
            } else {
                format!("id={}", v["versionId"].as_i64().unwrap_or(0))
            };
            let status = v["appStatus"].as_str().unwrap_or("unknown");
            format!("  v{} — {}", ver, status)
        })
        .collect();
    println!("{}:\n{}", package, lines.join("\n"));
    Ok(())
}

pub fn discard(package: &str) -> Result<()> {
    let state: DraftState = draft::load("rustore", package)?;
    let token = get_token()?;
    delete_version(package, state.version_id, &token);
    draft::delete("rustore", package);
    println!("RuStore draft deleted for {}", package);
    Ok(())
}

// ── Helpers ──────────────────────────────────────────────────────────────────

fn upsert_note(notes: &mut Vec<Note>, language: &str, text: &str) {
    if let Some(idx) = notes.iter().position(|n| n.language == language) {
        notes[idx].text = text.to_string();
    } else {
        notes.push(Note { language: language.to_string(), text: text.to_string() });
    }
}
